### DISCUSSION TITLE
latest/utils/video/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#1303] [Docs]
# latest/utils/video/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/latest/utils/video/

**Comment by AdelShousha** [comment]
can get_video_frames_generator take rtsp link as input?. if not what changes should I make and what should I watch out for when using the library?. I'm creating a license plate detection code that uses a real time rtsp link of a car parking entrance as input and the idea is that I want to give a tracking id for each entering car and proccess each frame that the car appears in taking it's license plate making a recognition then after the car disappears I'm going to take the recognition with the highest score and store it

**Reply by onuralpszr** [reply]
No you cannot, but you don't need frames generator for rtsp to get frame because rtsp is in loop and each frame already image itself (numpy) so you can just directly use it. Or you can use roboflow inference and get frame again then annotate it.

**Reply by LinasKo** [reply]
Hi @AdelShousha :wave:

The typical way to interact with RTSP streams is via the `InferencePipeline`. Here's an example:
https://blog.roboflow.com/computer-vision-rtsp-camera/

Here's the official documentation:
https://inference.roboflow.com/using_inference/inference_pipeline/

Hope this helps!

**Comment by kumarabhisek91** [comment]
While tracking object through bytetracker we are facing issue of id swapping. If 2 person cross each other in a video frame, there tracking id is getting swapped. How to tackle this issue. Whether bytetracker has any in-build method to tackle this issue.

**Reply by SkalskiP** [reply]
Unfortunately, ByteTrack only uses information about the position of the bounding box and does not take into account the appearance of the object. Therefore, I'm afraid it might be too weak to solve your problem. We are currently working on integrating Supervision with other trackers.

**Comment by ReyJust** [comment]
Is it possible to integrate tqdm with the `process_video` method?

**Reply by SkalskiP** [reply]
I was thinking about adding `tqdm` as a project dependency. Alternatively, you can use more verbose API and inject `tqdm` on your own.

```python
from tqdm import tqdm
import numpy as np
import supervision as sv


def callback(frame: np.ndarray, index: int) -> np.ndarray:
    # your code


source_video_info = VideoInfo.from_video_path(video_path=source_path)
with VideoSink(target_path=target_path, video_info=source_video_info) as sink:
    for index, frame in enumerate(
        tqdm(get_video_frames_generator(source_path=source_path), total=source_video_info.total_frames)
    ):
        result_frame = callback(frame, index)
        sink.write_frame(frame=result_frame)
```

**Reply by tonmoy-TS** [reply]
Try this:
```
from tqdm import tqdm
import numpy as np
import supervision as sv
# Your code

video_info = sv.VideoInfo.from_video_path(SOURCE_VIDEO_PATH)

def callback(frame: np.ndarray, index: int) -> np.ndarray:
    # Your code
    pbar.update(1) 

    return annotated_frame

with tqdm(total=video_info.total_frames, desc="Processing Video", unit="frame") as pbar:
    sv.process_video(
        source_path=SOURCE_VIDEO_PATH,
        target_path=TARGET_VIDEO_PATH,
        callback=lambda frame, index: callback(frame, index)
    )
```

**Comment by Muhammad-237** [comment]
how can we connect it with cctv stream can anyone help me and Gove a code snippet

**Comment by tonmoy-TS** [comment]
I am wondering how to use 'stride' to skip frames while video processing ?
Previously, I had been using the 'stride' argument inside the sv.process_video :
```
# Sample code
model = YOLO(<SOURCE_MODEL_PATH>)
...
video_info = sv.VideoInfo.from_video_path(SOURCE_VIDEO_PATH)
generator = sv.get_video_frames_generator(SOURCE_VIDEO_PATH)
iterator = iter(generator)
frame = next(iterator)
...
results = model(frame)[0]
bounding_box_annotator = sv.BoxAnnotator(thickness=3)

def callback(frame: np.ndarray, index: int, fps: int) -> np.ndarray:
    detections = sv.Detections.from_ultralytics(results)
    frame = bounding_box_annotator.annotate(scene=annotated_frame, detections=detections)
    return frame

sv.process_video(
    source_path=SOURCE_VIDEO_PATH,
    target_path=TARGET_VIDEO_PATH,
    callback=lambda frame, index: callback(frame, index, fps=fps),
    stride=10
    )
```

But now, I guess in the 0.24.0 version the 'stride' argument has been removed from sv.process_video, and used as argument for the  sv.get_video_frames_generator. So what would be the correct way to process video with skipped frames using stride ?

**Reply by LinasKo** [reply]
Hi @tonmoy-TS :wave: 

I looked into the docs history and the changelog. I can't find a case where `sv.process_video` ever had a stride argument. It cannot skip any of the frames, though we'd welcome a PR adding the feature :wink:

The function that has a stride is `get_video_frames_generator`. To replicate the functionality of `process_video`, you may use `get_video_frames_generator` together with `VideoSink`.

Check out the docs here: https://supervision.roboflow.com/develop/utils/video/#supervision.utils.video.VideoSink

**Reply by tonmoy-TS** [reply]
Hello, @LinasKo , you are right. `processVideo` never had a '`stride`' argument. I don't remember exactly if I had modified the code to make it work earlier, but I can't find the changes now. Anyway, I was able to skip frames using '`stride`' inside `get_video_frames_generator`, and then `VideoSink` (also, tqdm to track the progress). Here's how to implement it in case anyone is wondering:

```
# Sample code
model = YOLO(<SOURCE_MODEL_PATH>)
...
stride =15
video_info = sv.VideoInfo.from_video_path(SOURCE_VIDEO_PATH)
fps = video_info.fps

generator = sv.get_video_frames_generator(SOURCE_VIDEO_PATH, stride=stride)
iterator = iter(generator)
frame = next(iterator)
...
results = model(frame)[0]
bounding_box_annotator = sv.BoxAnnotator(thickness=3)

def callback(frame: np.ndarray, index: int, fps: int) -> np.ndarray:
    detections = sv.Detections.from_ultralytics(results)
    ....
    annotated_frame = bounding_box_annotator.annotate(scene=annotated_frame, detections=detections)

    #pbar.update(1) 

    return annotated_frame

'''
# Using sv.processVideo, (stride not applied):
with tqdm(total=video_info.total_frames, desc="Processing Video", unit="frame") as pbar:
    sv.process_video(
        source_path=SOURCE_VIDEO_PATH,
        target_path=TARGET_VIDEO_PATH,
        callback=lambda frame, index: callback(frame, index, fps=fps),
    )
'''

# Using VideoSink:
with sv.VideoSink(target_path=TARGET_VIDEO_PATH, video_info=video_info) as sink:
    for index, frame in enumerate(tqdm(generator, desc="Processing Video", unit="frame", total=video_info.total_frames // stride)):
        annotated_frame = callback(frame, index, fps)
        sink.write_frame(annotated_frame)

 ```

Thanks.

**Reply by LinasKo** [reply]
Thanks for the follow-up @tonmoy-TS ! Great to see you solved it, and I'm sure it'll help someone else (or us) in the future
